<a href="https://colab.research.google.com/github/NatalyaPetykhova/Practicum_projects/blob/main/%D0%9F%D1%80%D0%BEe%D0%BA%D1%827_%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7_%D0%B8%D0%BD%D0%B4%D1%83%D1%81%D1%82%D1%80%D0%B8%D0%B8_%D0%B8%D0%B3%D1%80.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Привет еще раз. Спасибо, что доделала работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация.

</div>

<div style="border: 2px solid black; padding: 10px; border-radius: 5px; background-color: #f0f0f0;">
    
Привет, Наталья.

Меня зовут Людмила, и я буду проводить ревью твоего проекта.

Предлагаю перейти на «ты», если не против.

Но если это не удобно — дай знать, и мы перейдем на «вы». Моя главная задача — не находить сделанные тобою ошибки, а передать свой опыт и помочь тебе стать аналитиком данных.

Мои комментарии будут в ячейках «Комментарий от ревьюера».

---    
<div class="alert alert-block alert-danger">
    
Критическое замечание, которое следует исправить.
</div>
    
<div class="alert alert-block alert-warning">
    
Замечание\совет на будущее.
</div>
    
<div class="alert alert-block alert-success">
    
Когда всё сделано правильно.
</div>
    
---        
    
Если ты вносишь изменения в проект по моим рекомендациям, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь свои комментарии заметным цветом или поместишь

<div class="alert alert-info">
    
в такую ячейку
</div>

Давай при работе с проектом мы будем общаться с тобой в диалоге, поэтому не удаляй и не перемещай мои комментарии, это поможет мне сделать повторную  проверку проекта быстрее.
___

# Анализ индустрии игр в начале XXI века

- Автор: Петухова Наталья
- Дата: 02.12.2024

### Цели и задачи проекта

Цель: Изучить развитие игровой индустрии. Сделать обзор игровых платформ, изучить объёмы продаж игр разных жанров и региональные предпочтения игроков.

Задачи: Познакомиться с данными, проверить их корректность и провести предобработку, получив необходимый срез данных.

### Описание данных

В проекте будут использованы данные датасета `/datasets/new_games.csv` с таким описанием:

- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.



### Содержимое проекта

1. Знакомство с данными
2. Проверка ошибок в данных и их предобработка
3. Фильтрация данных
4. Категоризация данных
5. Итоговый вывод


---

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>

Описание проекта составлено хорошо. Здорово, что есть чёткая формулировка цели, разделение задач по ключевым аспектам и описание датасета.

</div>

## Загрузка и знакомство с данными

1. Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.
2. Познакомьтесь с данными: выведите первые строки и результат метода `.info()`.
3. Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
4. Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

In [ ]:
# Импортируем библиотеку pandas
import pandas as pd
import numpy as np

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>

Хорошо, что импорты вынесены в отдельную ячейку. Это позволяет сразу увидеть, какие библиотеки и модули используются в проекте. Такой подход делает код более читаемым и понятным для других разработчиков и проверяющих.
</div>

In [ ]:
# Выгружаем данные из датасета new_games.csv в датафрейм new_games
new_games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [ ]:
# Выводим первые стороки таблицы для знакомства с данными
new_games.head(15)

Name Platform  Year of Release         Genre  \
0                    Wii Sports      Wii           2006.0        Sports   
1             Super Mario Bros.      NES           1985.0      Platform   
2                Mario Kart Wii      Wii           2008.0        Racing   
3             Wii Sports Resort      Wii           2009.0        Sports   
4      Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                        Tetris       GB           1989.0        Puzzle   
6         New Super Mario Bros.       DS           2006.0      Platform   
7                      Wii Play      Wii           2006.0          Misc   
8     New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                     Duck Hunt      NES           1984.0       Shooter   
10                   Nintendogs       DS           2005.0    Simulation   
11                Mario Kart DS       DS           2005.0        Racing   
12  Pokemon Gold/Pokemon Silver       GB           1999.0  Role-Playing   
13                      Wii Fit      Wii           2007.0        Sports   
14           Kinect Adventures!     X360           2010.0          Misc   

    NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0      41.36    28.96     3.77         8.45          76.0          8      E  
1      29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2      15.68    12.76     3.79         3.29          82.0        8.3      E  
3      15.61    10.93     3.28         2.95          80.0          8      E  
4      11.27     8.89    10.22         1.00           NaN        NaN    NaN  
5      23.20     2.26     4.22         0.58           NaN        NaN    NaN  
6      11.28     9.14      6.5         2.88          89.0        8.5      E  
7      13.96     9.18     2.93         2.84          58.0        6.6      E  
8      14.44     6.94      4.7         2.24          87.0        8.4      E  
9      26.93     0.63     0.28         0.47           NaN        NaN    NaN  
10      9.05    10.95     1.93         2.74           NaN        NaN    NaN  
11      9.71     7.47     4.13         1.90          91.0        8.6      E  
12      9.00     6.18      7.2         0.71           NaN        NaN    NaN  
13      8.92     8.03      3.6         2.15          80.0        7.7      E  
14     15.00     4.89     0.24         1.69          61.0        6.3      E

In [ ]:
# Выводим информацию о датафрейме
new_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>

Основная информация о таблице выведена, супер.
</div>

Датасет new_games.csv сожержит 11 столбцов и 16956 строк в которых представлена информация о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.

Изучив первые строки таблицы, в данных обнаружены пропуски в столбцах `Name`, `Year of Release`, `Genre`, `Critic Score`, `User Score`, `Rating`. Названия столбцов и данные в них соответствуют описанию, но названия столбцов представлены в неудобном для работы виде, нужно привести их к стилю snake_case.

Изучим типы данных и их корректность:
- **Числовые значения с плавающей точкой (float64).** Четыре столбца имеют тип данных `float64`:
    - Столбец `Year of Release`, содержит информацию о дате выпуска игры и представлен типом `float64`. Это не верное решение, так, как дата должна быть представлена типом `int64`.
    - Столбец `NA sales`, сдержит информацию о продажах в Северной Америке (в миллионах проданных копий), а столбец `Other sales`, продажи в других странах (в миллионах проданных копий). Столбцы представлены типом `float64`, это верное решение.
    - Столбец `Critic Score`, содержит информацию об оценке критиков (от 0 до 100). Данный столбец представлен типом `float64`, это верное решение.
- **Строковые данные (object).** Семь столбцов имеют тип данных `object`:
    - `Name`, `Platform`, `Genre` и `Rating` содержат строковую информацию (название игры, название платформы, жанр и рейтинг организации ESRB) и представлены типом `object`, что логично для текстовых данных. Здесь тип данных `object` подходит.
    - `EU sales` и `JP sales` хранят информацию о продажах в Европе и продажах в Японии (в миллионах проданных копий). Для таких данных рекомендуется использовать тип `float64`.
    - Столбец `User Score` хранит информацию об оценке пользователей (от 0 до 10) и представлен типом `object`. Это не верное решение и данный столбец нужно привести к типу `float64`.

После анализа типов данных видно, что некоторые столбцы представлены некорректно и нужно выполнить преобразование типов данных.


<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>

Получился очень подробный и хорошо структурированный вывод, что демонстрирует внимательность к деталям.
    
</div>

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b>

Вывод можно немного дополнить для большей ясности и глубины анализа:
    
* указать, что столбцы необходимо переименовать, приведя названия к snake_case;
* предположить, почему столбцы EU sales, JP sales и User Score не распознаются как числовые.
    
</div>

<div class="alert alert-info">
Добрый день Людмила! Спасибо тебе за комментарии)

Информация о том, что столбцы представлены в неудобном для работы виде и их нужно привести к змеиному регистру в выводе есть, во втором абзаце.

Предположения о том, почему столбцы не распознаются как числовые, я написала в пункте 2.2 Типы данных
</div>

---

## Проверка ошибок в данных и их предобработка


### Названия, или метки, столбцов датафрейма:

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [ ]:
# Выводим названия столбцов, для того чтобы проверить стиль написания
columns_names = new_games.columns
display(columns_names)

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [ ]:
# Приводим названия колонок в правильный формат
new_games.columns = columns_names.str.lower().str.replace(" ", "_")
new_games.head(20)

name platform  year_of_release  \
0                                     Wii Sports      Wii           2006.0   
1                              Super Mario Bros.      NES           1985.0   
2                                 Mario Kart Wii      Wii           2008.0   
3                              Wii Sports Resort      Wii           2009.0   
4                       Pokemon Red/Pokemon Blue       GB           1996.0   
5                                         Tetris       GB           1989.0   
6                          New Super Mario Bros.       DS           2006.0   
7                                       Wii Play      Wii           2006.0   
8                      New Super Mario Bros. Wii      Wii           2009.0   
9                                      Duck Hunt      NES           1984.0   
10                                    Nintendogs       DS           2005.0   
11                                 Mario Kart DS       DS           2005.0   
12                   Pokemon Gold/Pokemon Silver       GB           1999.0   
13                                       Wii Fit      Wii           2007.0   
14                            Kinect Adventures!     X360           2010.0   
15                                  Wii Fit Plus      Wii           2009.0   
16                            Grand Theft Auto V      PS3           2013.0   
17                 Grand Theft Auto: San Andreas      PS2           2004.0   
18                             Super Mario World     SNES           1990.0   
19  Brain Age: Train Your Brain in Minutes a Day       DS           2005.0   

           genre  na_sales eu_sales jp_sales  other_sales  critic_score  \
0         Sports     41.36    28.96     3.77         8.45          76.0   
1       Platform     29.08     3.58     6.81         0.77           NaN   
2         Racing     15.68    12.76     3.79         3.29          82.0   
3         Sports     15.61    10.93     3.28         2.95          80.0   
4   Role-Playing     11.27     8.89    10.22         1.00           NaN   
5         Puzzle     23.20     2.26     4.22         0.58           NaN   
6       Platform     11.28     9.14      6.5         2.88          89.0   
7           Misc     13.96     9.18     2.93         2.84          58.0   
8       Platform     14.44     6.94      4.7         2.24          87.0   
9        Shooter     26.93     0.63     0.28         0.47           NaN   
10    Simulation      9.05    10.95     1.93         2.74           NaN   
11        Racing      9.71     7.47     4.13         1.90          91.0   
12  Role-Playing      9.00     6.18      7.2         0.71           NaN   
13        Sports      8.92     8.03      3.6         2.15          80.0   
14          Misc     15.00     4.89     0.24         1.69          61.0   
15        Sports      9.01     8.49     2.53         1.77          80.0   
16        Action      7.02     9.09     0.98         3.96          97.0   
17        Action      9.43      0.4     0.41        10.57          95.0   
18      Platform     12.78     3.75     3.54         0.55           NaN   
19          Misc      4.74      9.2     4.16         2.04          77.0   

   user_score rating  
0           8      E  
1         NaN    NaN  
2         8.3      E  
3           8      E  
4         NaN    NaN  
5         NaN    NaN  
6         8.5      E  
7         6.6      E  
8         8.4      E  
9         NaN    NaN  
10        NaN    NaN  
11        8.6      E  
12        NaN    NaN  
13        7.7      E  
14        6.3      E  
15        7.4      E  
16        8.2      M  
17          9      M  
18        NaN    NaN  
19        7.9      E

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Отлично, столбцы выведены на экран и переименованы.
</div>

Для более удобной работы приводим названия столбцов к стилю snake case

### Типы данных:

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу int64. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.
- В числовых столбцах могут встретиться строковые значения, например, `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

В данных встречаются столбцы с некорректными типами данных и пропусками, выполним преобразование типов.

Типы данных при выгрузке назначаются автоматически, поэтому некоторым столбцам, по ошибке назначились некорректные типы данных.

Столбцы `EU sales`, `JP sales` и `User Score` имеют тип данных `object` возможно потому, что в них содержатся строковые значения.

In [ ]:
# Перед преобразованием выведем уникальные значения в столбцах, для того, чтобы проверить, нет ли в числовых столбцах текстовых значений
unique_values_year_of_release = new_games['year_of_release'].unique()
unique_values_eu_sales = new_games['eu_sales'].unique()
unique_values_jp_sales = new_games['jp_sales'].unique()
unique_values_user_score = new_games['user_score'].unique()
display(unique_values_year_of_release, unique_values_eu_sales, unique_values_jp_sales, unique_values_user_score)

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

В столбцах `eu_sales`, `jp_sales`, `user_score` найдены текстовые значения. Поэтому для преобразования используем метод `pd.to_numeric()` с параметром `errors='coerce'`. Этот параметр позволяет обрабатывать ошибки, возникающие при попытке преобразования значений, которые не могут быть корректно интерпретированы. В данном случае все значения, которые не могут быть преобразованы, будут заменены на NaN.

In [ ]:
# Преобразуем тип данных столбца 'year_of_release' к типу Int64
new_games['year_of_release'] = new_games['year_of_release'].astype('Int64')

In [ ]:
# Выводим минимальную и максимальную дату, для того чтобы посмотреть в каком временном промежутке у нас данные
display(new_games['year_of_release'].min())
display(new_games['year_of_release'].max())

1980

2016

<div class="alert alert-block alert-danger">
    
<b>Комментарий ревьюера:</b>
    
Перед преобразованием типов данных стоит вывести уникальные значения в столбцах. Так мы сможем увидеть текстовые значения, из-за которых столбец не распознается как числовой, что сделает наше преобразование более обоснованным.
    
</div>

<div class="alert alert-info">
Исправила. Уникальные значения в столбцах вывела, добавила вывод.
        
    В столбцах eu_sales, jp_sales, user_score найдены текстовые значения. Поэтому для преобразования используем метод    pd.to_numeric() с параметром errors='coerce'. Этот параметр позволяет обрабатывать ошибки, возникающие при попытке   преобразования значений, которые не могут быть корректно интерпретированы. В данном случае все значения, которые не могут быть преобразованы, будут заменены на NaN.
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Всё верно, молодец, что добавила вывод.

</div>

In [ ]:
# Преобразуем тип данных столбцов 'eu_sales', 'jp_sales' и 'user_score' к типу float64
new_games['eu_sales'] = pd.to_numeric(new_games['eu_sales'], errors = 'coerce')

new_games['jp_sales'] = pd.to_numeric(new_games['jp_sales'], errors = 'coerce')

new_games['user_score'] = pd.to_numeric(new_games['user_score'], errors = 'coerce')

In [ ]:
# Проверяем, выполнилось ли преобразование
new_games.dtypes

name                object
platform            object
year_of_release      Int64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Хорошо, что ты проверяешь результат после преобразования.
</div>

Преобразование 4 столбцов прошло успешно. Теперь все столбцы имеют правильный тип данных.

### Наличие пропусков в данных:

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.
- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.
- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [ ]:
# Посчитаем количество строк с пропусками в каждом столбце
new_games.isna().sum().sort_values(ascending = False)

user_score         9268
critic_score       8714
rating             6871
year_of_release     275
eu_sales              6
jp_sales              4
name                  2
genre                 2
platform              0
na_sales              0
other_sales           0
dtype: int64

In [ ]:
# Посчитаем процент строк с пропусками
new_games.isna().mean().sort_values(ascending = False)* 100

user_score         54.659118
critic_score       51.391838
rating             40.522529
year_of_release     1.621845
eu_sales            0.035386
jp_sales            0.023590
name                0.011795
genre               0.011795
platform            0.000000
na_sales            0.000000
other_sales         0.000000
dtype: float64

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Хорошо, что ты отсортировала результаты по убыванию Это правильно и помогает сразу увидеть, в каких столбцах больше всего пропусков.
</div>

В столбце `critic_score` (оценка критиков (от 0 до 100)), 9268 строк с пропущенными значениями, это 54,6%. Пропуски могли пойвится в результате того, что критики не захотели поставить оценку или еще не играли в игру, и поэтому оценки нет. Пропуски можно можно заполнить значением, которое не попадает в диапазон от 0 - 100.

В столбце `rating` (рейтинг организации ESRB), 8714 строка с пропускми, это 51,4%. Пропуски могли появится в результате того, что рейтинг не был присвоен или для старых игр был утерян. Данные пропуски можно заполнить значением-индикатор, чтобы оно не попало в диапазон, например `unknown`.

В столбце `user_score` (оценка пользователей (от 0 до 10)), 6871 строки с пропущенными значениями, это 40,5%. Пропуски могли пойвится в результате того, что пользователи не захотели поставить оценку, забыли поставить оценку или еще не закончили игру, и поэтому оценки нет. Пропуски можно можно заполнить значением, которое не попадает в диапазон от 0 - 10.

В столбце `year_of_release` (год выпуска игры), 275 пропусков, это всего 1,62%. Данные могли быть утеряны или не заполнены, в результате чего появились пропуски. Год релиза — ключевой параметр для анализа временных трендов, таких как популярность жанров, платформ и динамика продаж. Поэтому чтобы избежать возможных ошибок в дальнейшем анализе и упростить работу с данными удалим эти строки.

В столбце `eu_sales` (продажи в Европе), 6 строк с пропусками, это меньше 1%, поэтому пропуски можно заполнить средним значением.

В столбце `jp_sales` (продажи в Японии), 4 строки с пропусками, это меньше 1%, поэтому пропуски можно заполнить средним значением.

В столбце `name` (название игры), 2 строки с пропусками, это меньше 1%, поэтому пропуски можно удалить.

В столбце `genre` (жанр игры), 2 строки с пропусками, это меньше 1%, поэтому пропуски можно удалить.



In [ ]:
# Заполняем пропуски значением-индикатор
new_games['critic_score'] = new_games['critic_score'].fillna(-1)
new_games['rating'] = new_games['rating'].fillna('unknown')
new_games['user_score'] = new_games['user_score'].fillna(-1)


<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b>
    
Столбец с рейтингом у нас текстовый, поэтому правильнее будет заполнить его соответствующей заглушой (например unknown).
    
Год релиза — ключевой параметр для анализа временных трендов, таких как популярность жанров, платформ и динамика продаж. Пропуски в этом столбце составляют всего 1,62% от общего числа строк, поэтому проще удалить данные строки, чтобы избежать возможных ошибок в дальнейшем анализе и упростить работу с данными.
</div>

<div class="alert alert-info">
Исправила. Столбец 'rating' заполнила заглушкой 'unknown'
Пропуски в столбце 'year_of_release' удалила.   
Информацию в выводах скорректировала.    
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Хорошо, что ты не забываешь корректировать выводы.
</div>

In [ ]:
# Заполняем пропуски в столбце 'eu_sales' средним значением расчитанным для каждой группы
def mean_group_eu_sales(row):
    if pd.isna(row['eu_sales']):
        group = new_games[(new_games['platform'] == row['platform']) &
                   (new_games['year_of_release'] == row['year_of_release'])]
        return group['eu_sales'].mean()
    else:
        return row['eu_sales']

new_games['eu_sales'] = new_games.apply(mean_group_eu_sales, axis=1)
display(new_games.head(10))

name platform  year_of_release         genre  \
0                 Wii Sports      Wii             2006        Sports   
1          Super Mario Bros.      NES             1985      Platform   
2             Mario Kart Wii      Wii             2008        Racing   
3          Wii Sports Resort      Wii             2009        Sports   
4   Pokemon Red/Pokemon Blue       GB             1996  Role-Playing   
5                     Tetris       GB             1989        Puzzle   
6      New Super Mario Bros.       DS             2006      Platform   
7                   Wii Play      Wii             2006          Misc   
8  New Super Mario Bros. Wii      Wii             2009      Platform   
9                  Duck Hunt      NES             1984       Shooter   

   na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
0     41.36     28.96      3.77         8.45          76.0         8.0   
1     29.08      3.58      6.81         0.77          -1.0        -1.0   
2     15.68     12.76      3.79         3.29          82.0         8.3   
3     15.61     10.93      3.28         2.95          80.0         8.0   
4     11.27      8.89     10.22         1.00          -1.0        -1.0   
5     23.20      2.26      4.22         0.58          -1.0        -1.0   
6     11.28      9.14      6.50         2.88          89.0         8.5   
7     13.96      9.18      2.93         2.84          58.0         6.6   
8     14.44      6.94      4.70         2.24          87.0         8.4   
9     26.93      0.63      0.28         0.47          -1.0        -1.0   

    rating  
0        E  
1  unknown  
2        E  
3        E  
4  unknown  
5  unknown  
6        E  
7        E  
8        E  
9  unknown

In [ ]:
# Заполняем пропуски в столбце 'jp_sales' средним значением расчитанным для каждой группы
def mean_group_jp_sales(row):
    if pd.isna(row['jp_sales']):
        group = new_games[(new_games['platform'] == row['platform']) &
                   (new_games['year_of_release'] == row['year_of_release'])]
        return group['jp_sales'].mean()
    else:
        return row['jp_sales']

new_games['jp_sales'] = new_games.apply(mean_group_jp_sales, axis=1)
display(new_games.head(10))

name platform  year_of_release         genre  \
0                 Wii Sports      Wii             2006        Sports   
1          Super Mario Bros.      NES             1985      Platform   
2             Mario Kart Wii      Wii             2008        Racing   
3          Wii Sports Resort      Wii             2009        Sports   
4   Pokemon Red/Pokemon Blue       GB             1996  Role-Playing   
5                     Tetris       GB             1989        Puzzle   
6      New Super Mario Bros.       DS             2006      Platform   
7                   Wii Play      Wii             2006          Misc   
8  New Super Mario Bros. Wii      Wii             2009      Platform   
9                  Duck Hunt      NES             1984       Shooter   

   na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
0     41.36     28.96      3.77         8.45          76.0         8.0   
1     29.08      3.58      6.81         0.77          -1.0        -1.0   
2     15.68     12.76      3.79         3.29          82.0         8.3   
3     15.61     10.93      3.28         2.95          80.0         8.0   
4     11.27      8.89     10.22         1.00          -1.0        -1.0   
5     23.20      2.26      4.22         0.58          -1.0        -1.0   
6     11.28      9.14      6.50         2.88          89.0         8.5   
7     13.96      9.18      2.93         2.84          58.0         6.6   
8     14.44      6.94      4.70         2.24          87.0         8.4   
9     26.93      0.63      0.28         0.47          -1.0        -1.0   

    rating  
0        E  
1  unknown  
2        E  
3        E  
4  unknown  
5  unknown  
6        E  
7        E  
8        E  
9  unknown

<div class="alert alert-block alert-danger">
    
<b>Комментарий ревьюера:</b>
    
В текущем коде пропуски в столбцах jp_sales и eu_sales заменяются средним значением по всему столбцу. Однако в задании требуется заменить пропуски на среднее значение, рассчитанное для каждой группы, определяемой комбинацией платформы (platform) и года выпуска игры (year_of_release). Такой подход позволяет учитывать особенности продаж для каждой платформы в конкретные годы, что важно для корректного анализа, например, при изучении динамики продаж или популярности платформ.

</div>

<div class="alert alert-info">
Исправила. Заменила пропуски в столбцах jp_sales и eu_sales на среднее значение расчитанное для каждой группы, с помощью функции и метода apply()   
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Да, это то что нас просят сделать в задании.

</div>

In [ ]:
# Удаляем пропуски в столбцах 'name', 'genre' и 'year_of_release'
new_games = new_games.dropna(subset=['name', 'genre', 'year_of_release'])

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Да, тут всего 2 пропуска, поэтому можно просто удалить данные строки.
</div>

In [ ]:
# Проверяем результат работы
new_games.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

Теперь пропусков в данных нет.


### Явные и неявные дубликаты в данных:

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.
- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.
- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

В процессе подготовки данных вы могли удалять данные, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях. После проведения предобработки данных напишите общий промежуточный вывод.

In [ ]:
# Проверим, есть ли дубликаты в названиях жанров игр
unique_values = new_games['genre'].unique()
display(unique_values)

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

Жанры игр написаны в разном регистре, приведем все названия жанров к нижнему регистру

In [ ]:
# Приводим названия жанров к нижнему регистру
new_games['genre'] = new_games['genre'].str.lower()


<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Хороший подход к приведению жанров к одному регистру, что позволяет избежать неявных дубликатов.
</div>

In [ ]:
# Проверим, есть ли дубликаты в названиях игровых платформ
unique_values = new_games['platform'].unique()
display(unique_values)

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

Среди названий платформ дубликатов нет, все значения уникальны

Консоли и Портативные Системы
- Wii: Nintendo Wii
- NES: развлекательная система Nintendo
- GB: Game Boy
- DS: Nintendo DS
- X360: Xbox 360
- PS3: PlayStation 3
- PS2: PlayStation 2
- SNES: развлекательная система Super Nintendo
- GBA: Game Boy Advance
- PS4: PlayStation 4
- 3DS: Nintendo 3DS
- N64: Nintendo 64
- PS: PlayStation (оригинальная)
- XB: Xbox (оригинальная)
- PC: персональный компьютер (не является конкретной консолью, а скорее общим термином для компьютеров)
- 2600: Atari 2600
- PSP: PlayStation Portable
- XOne: Xbox One
- WIIIU: Wii U
- GC: Nintendo GameCube
- GEN: Sega Genesis (также известна как Sega Mega Drive)
- DC: Sega Dreamcast
- PSV: PlayStation Vita
- SAT: Sega Saturn
- SCD: Компакт-диск Sega
- WS: WonderSwan
- NG: Neo Geo
- TG16: TurboGrafx-16 (также известная как PC Engine)

Другие Платформы
- 3DO: интерактивный мультиплеер 3DO
- GG: Игровое оборудование Sega
- PCFX: PC-FX

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Здорово, что ты расшифровала названия всех платформ. Это демонстрирует внимательность и глубокое понимание данных.
</div>

In [ ]:
# Проверим, есть дубликаты в данных о рейтинге
unique_values = new_games['rating'].unique()
display(unique_values)

array(['E', 'unknown', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'],
      dtype=object)

Дубликатов в данных рейтинга нет. Но есть значение которого нет в расшифровках аббревиатуры рейтингов - К-А.

По данным интернета этот рейтинг использовался до 1998 года, когда его переименовали в E.

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Всё верно, молодец, что используешь дополнительную информацию для выполнения проекта.
</div>

In [ ]:
# Заменим значение рейтинга на актуальную информацию
new_games['rating'] = new_games['rating'].replace('K-A', 'E')

unique_values = new_games['rating'].unique()
display(unique_values)

array(['E', 'unknown', 'M', 'T', 'E10+', 'AO', 'EC', 'RP'], dtype=object)

In [ ]:
# Проверим, есть дубликаты в столбце год выхода игры
unique_values = new_games['year_of_release'].unique()
display(unique_values)

<IntegerArray>
[2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010, 2013, 2004,
 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014, 1992, 1997, 1993, 1994,
 1982, 2016, 2003, 1986, 2000, 1995, 1991, 1981, 1987, 1980, 1983]
Length: 37, dtype: Int64

Дубликатов в данных о годе выхода игр нет.

Неявные дубликаты выявили, теперь проверим данные на наличие явных дубликатов.

In [ ]:
# Отсортируем датафрейм по всем столбцам в порядке возрастания — так будет удобнее находить дубликаты
new_games_sorted = new_games.sort_values(by=new_games.columns.tolist())
display(new_games_sorted)

name platform  year_of_release         genre  \
15191                Beyblade Burst      3DS             2016  role-playing   
15192                Beyblade Burst      3DS             2016  role-playing   
1086              Fire Emblem Fates      3DS             2015  role-playing   
3394           Frozen: Olaf's Quest      3DS             2013      platform   
3906           Frozen: Olaf's Quest       DS             2013      platform   
...                             ...      ...              ...           ...   
16731    thinkSMART: Chess for Kids       DS             2011          misc   
647                    uDraw Studio      Wii             2010          misc   
8397   uDraw Studio: Instant Artist      Wii             2011          misc   
15836  uDraw Studio: Instant Artist     X360             2011          misc   
9260   ¡Shin Chan Flipa en colores!       DS             2007      platform   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
15191      0.00      0.00      0.03         0.00          -1.0        -1.0   
15192      0.00      0.00      0.03         0.00          -1.0        -1.0   
1086       0.81      0.23      0.52         0.11          -1.0        -1.0   
3394       0.27      0.27      0.00         0.05          -1.0        -1.0   
3906       0.21      0.26      0.00         0.04          -1.0        -1.0   
...         ...       ...       ...          ...           ...         ...   
16731      0.01      0.00      0.00         0.00          -1.0        -1.0   
647        1.65      0.57      0.00         0.20          71.0        -1.0   
8397       0.06      0.09      0.00         0.02          -1.0        -1.0   
15836      0.01      0.01      0.00         0.00          54.0         5.7   
9260       0.00      0.00      0.14         0.00          -1.0        -1.0   

        rating  
15191  unknown  
15192  unknown  
1086   unknown  
3394   unknown  
3906   unknown  
...        ...  
16731        E  
647          E  
8397         E  
15836        E  
9260   unknown  

[16679 rows x 11 columns]

In [ ]:
# Посчитаем общее количество дубликатов
duplicates = new_games.duplicated().sum()
display(duplicates)

235

Теперь нужно удалить дубликаты во всем датафрейме

In [ ]:
# Сохраняем количество строк до удаления дубликатов
initial_row_count = new_games.shape[0]

In [ ]:
# Удаляем дубликаты
new_games_no_duplicates = new_games_sorted.drop_duplicates()

In [ ]:
# Сохраняем количество строк после удаления дубликатов
final_row_count = new_games_no_duplicates.shape[0]

In [ ]:
# Считаем количество удаленных строк
total_row_count = initial_row_count - final_row_count
perc_total_row_count = round((total_row_count / initial_row_count * 100), 2)
display(total_row_count, perc_total_row_count)

235

1.41

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Хорошо, что ты посчитала как количество, так и процент удаленных строк.
</div>

В результате проведения предобработки данных были выявлены и преобразованы неявные дубликаты в названиях жанров игр. С помощью приведения всех названий жанров к одинаковому регистру, дубликаты устранены.

Проверили наличие дубликатов в названиях игровых платформ. Среди названий платформ все значения уникальны.

Во время поиска дубликатов в данных о рейтингах выявлено значение которого нет в расшифровках аббревиатуры рейтингов - К-А.
По данным интернета этот рейтинг использовался до 1998 года, когда его переименовали в E. Значение рейтинга К-А заменили на более актаульное - Е.

Дубликатов в данных о годе выхода игр нет.

Так же в данных обнаружены и удалены явные дубликаты. Таких дубликатов 1,41%



---

## Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [ ]:
# Проводим фильтрацию по году выхода игр
new_games_actual = new_games[(new_games['year_of_release'] >= 2000) & (new_games['year_of_release'] <= 2013)]

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Отлично, фильтрация соответствует требованиям задания.
</div>

In [ ]:
# Сортируем данные по возрастанию
new_games_sort = new_games_actual.sort_values(by='year_of_release')
display(new_games_sort.head(10))

name platform  year_of_release  \
10435             FoxKids.com Micro Maniacs Racing       PS             2000   
3970                                 Final Fantasy       WS             2000   
370                    Rugrats in Paris: The Movie       PS             2000   
6352                   Evil Dead: Hail to the King       PS             2000   
2959         Mobile Suit Gundam: Journey to Jaburo      PS2             2000   
1279                                   Banjo-Tooie      N64             2000   
8742                          NCAA Final Four 2001       PS             2000   
1269                                  WWF No Mercy      N64             2000   
362                                 WWF SmackDown!       PS             2000   
10830  Sakura Wars 2: Kimi, Shinitamou koto Nakare       DC             2000   

              genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
10435        racing      0.06      0.04      0.00         0.01          -1.0   
3970   role-playing      0.00      0.00      0.51         0.00          -1.0   
370          action      1.96      1.33      0.00         0.23          -1.0   
6352      adventure      0.15      0.10      0.00         0.02          51.0   
2959     simulation      0.16      0.13      0.36         0.04          58.0   
1279       platform      0.82      0.36      0.25         0.06          -1.0   
8742         sports      0.09      0.06      0.00         0.01          55.0   
1269       fighting      1.20      0.27      0.02         0.02          -1.0   
362        fighting      2.01      1.35      0.06         0.16          -1.0   
10830     adventure      0.00      0.00      0.10         0.00          -1.0   

       user_score   rating  
10435        -1.0  unknown  
3970         -1.0  unknown  
370          -1.0  unknown  
6352          6.0        M  
2959          8.1        T  
1279         -1.0  unknown  
8742         -1.0        E  
1269         -1.0  unknown  
362          -1.0  unknown  
10830        -1.0  unknown

Дополнительно расчитаем количество проданных игр по годам. Это позволит выявить тенденции в объёме выпускаемых игр: были ли годы с пиками или спадами в количестве выпущенных игр, поможет понять, насколько данные сбалансированы и можно ли делать обобщения для всего периода.

In [ ]:
# Вычисляем общее количество проданных игр
new_games['count_games'] = new_games['na_sales'] + new_games['eu_sales'] + new_games['jp_sales'] + new_games['other_sales']
# Группируем по году выпуска игр
count_games_year = new_games.groupby('year_of_release')['count_games'].count()
# Выводим результат на экран отсортировав по убыванию для того , чтобы посмотреть, в какие годы было продано наибольшее количество игр
display(count_games_year.sort_values(ascending = False))

year_of_release
2009    1450
2008    1445
2010    1279
2007    1218
2011    1149
2006    1020
2005     950
2002     839
2003     789
2004     771
2012     670
2015     612
2014     584
2013     552
2016     507
2001     491
1998     384
2000     357
1999     341
1997     293
1996     267
1995     220
1994     121
1993      60
1981      46
1992      43
1991      42
1982      37
1986      22
1983      18
1989      17
1987      17
1990      16
1988      15
1984      14
1985      14
1980       9
Name: count_games, dtype: int64

В результате вычислений видим, что в период с 2006-2011 год продажи компьютерных игр были на пике, а в период с 1980-1993 количество продаж компьютерных игр было самым низким. После 2000 количество выпускаемых игр стало расти.

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b>
    
Здесь можно дополнительно подсчитать количество игр по годам. Это позволит выявить тенденции в объёме выпускаемых игр: были ли годы с пиками или спадами в количестве выпущенных игр, поможет понять, насколько данные сбалансированы и можно ли делать обобщения для всего периода.
</div>

<div class="alert alert-info">
Исправила. Добавила расчет количества проданных игр по годам. Написала вывод по данным расчетам.
    
    В результате вычислений видим, что в период с 2006-2011 год продажи компьютерных игр были на пике, а в период с 1980-1993г. количество продаж компьютерных игр было самым низким. После 2000г. количество выпускаемых игр стало расти.
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Да, действительно, здорово, что ты обратила на это внимание.

</div>

---

## Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).
- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).
- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b>
    
В данном пункте будет лучше использовать уже отфильтрованный по годам датасет, так как в результате нам необходимо предоставить таблицу, обработанную по всем перечисленным требованиям.
</div>

<div class="alert alert-info">
Исправила. В данном  пункте для сортировки игр по оценкам пользователей и по оценкам критиков, использовала датасет отфильтрованный ранее по годам, с 2000-2013г.
</div>

In [ ]:
# С помощью пользовательской функции разделяем все игры по оценкам пользователей на три категории.
def user_score_categories(score):
    if  0 <= score < 3:
        return 'Низкая оценка'
    elif 3 <= score < 8:
        return 'Средняя оценка'
    elif 8 <= score <= 10:
        return 'Высокая оценка'
    else:
        return 'Нет данных'

new_games_sort['user_score_categories'] = new_games_sort['user_score'].apply(user_score_categories)
display(new_games_sort)

name platform  \
10435                  FoxKids.com Micro Maniacs Racing       PS   
3970                                      Final Fantasy       WS   
370                         Rugrats in Paris: The Movie       PS   
6352                        Evil Dead: Hail to the King       PS   
2959              Mobile Suit Gundam: Journey to Jaburo      PS2   
...                                                 ...      ...   
7104                                       Game & Wario     WiiU   
4739                                           Terraria      PS3   
14006  J.League Pro Soccer Club o Tsukurou! 8 Euro Plus      PSP   
14033                          Date A Live: Rine Utopia      PS3   
13347                               Winning Post 7 2013      PS3   

       year_of_release         genre  na_sales  eu_sales  jp_sales  \
10435             2000        racing      0.06      0.04      0.00   
3970              2000  role-playing      0.00      0.00      0.51   
370               2000        action      1.96      1.33      0.00   
6352              2000     adventure      0.15      0.10      0.00   
2959              2000    simulation      0.16      0.13      0.36   
...                ...           ...       ...       ...       ...   
7104              2013          misc      0.05      0.06      0.12   
4739              2013        action      0.07      0.20      0.09   
14006             2013        sports      0.00      0.00      0.04   
14033             2013     adventure      0.00      0.00      0.04   
13347             2013        sports      0.00      0.00      0.05   

       other_sales  critic_score  user_score   rating user_score_categories  
10435         0.01          -1.0        -1.0  unknown            Нет данных  
3970          0.00          -1.0        -1.0  unknown            Нет данных  
370           0.23          -1.0        -1.0  unknown            Нет данных  
6352          0.02          51.0         6.0        M        Средняя оценка  
2959          0.04          58.0         8.1        T        Высокая оценка  
...            ...           ...         ...      ...                   ...  
7104          0.01          -1.0        -1.0  unknown            Нет данных  
4739          0.05          81.0         7.9        T        Средняя оценка  
14006         0.00          -1.0        -1.0  unknown            Нет данных  
14033         0.00          -1.0        -1.0  unknown            Нет данных  
13347         0.00          -1.0        -1.0  unknown            Нет данных  

[12980 rows x 12 columns]

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Здорово, что ты учитываешь категорию "Нет данных".
</div>

In [ ]:
# С помощью пользовательской функции разделяем все игры по оценкам критиков на три категории.
def critic_score_categories(score):
    if  0 <= score < 30:
        return 'Низкая оценка'
    elif 30 <= score < 80:
        return 'Средняя оценка'
    elif 80 <= score <= 100:
        return 'Высокая оценка'
    else:
        return 'Нет данных'

new_games_sort['critic_score_categories'] = new_games_sort['critic_score'].apply(critic_score_categories)
display(new_games_sort)

name platform  \
10435                  FoxKids.com Micro Maniacs Racing       PS   
3970                                      Final Fantasy       WS   
370                         Rugrats in Paris: The Movie       PS   
6352                        Evil Dead: Hail to the King       PS   
2959              Mobile Suit Gundam: Journey to Jaburo      PS2   
...                                                 ...      ...   
7104                                       Game & Wario     WiiU   
4739                                           Terraria      PS3   
14006  J.League Pro Soccer Club o Tsukurou! 8 Euro Plus      PSP   
14033                          Date A Live: Rine Utopia      PS3   
13347                               Winning Post 7 2013      PS3   

       year_of_release         genre  na_sales  eu_sales  jp_sales  \
10435             2000        racing      0.06      0.04      0.00   
3970              2000  role-playing      0.00      0.00      0.51   
370               2000        action      1.96      1.33      0.00   
6352              2000     adventure      0.15      0.10      0.00   
2959              2000    simulation      0.16      0.13      0.36   
...                ...           ...       ...       ...       ...   
7104              2013          misc      0.05      0.06      0.12   
4739              2013        action      0.07      0.20      0.09   
14006             2013        sports      0.00      0.00      0.04   
14033             2013     adventure      0.00      0.00      0.04   
13347             2013        sports      0.00      0.00      0.05   

       other_sales  critic_score  user_score   rating user_score_categories  \
10435         0.01          -1.0        -1.0  unknown            Нет данных   
3970          0.00          -1.0        -1.0  unknown            Нет данных   
370           0.23          -1.0        -1.0  unknown            Нет данных   
6352          0.02          51.0         6.0        M        Средняя оценка   
2959          0.04          58.0         8.1        T        Высокая оценка   
...            ...           ...         ...      ...                   ...   
7104          0.01          -1.0        -1.0  unknown            Нет данных   
4739          0.05          81.0         7.9        T        Средняя оценка   
14006         0.00          -1.0        -1.0  unknown            Нет данных   
14033         0.00          -1.0        -1.0  unknown            Нет данных   
13347         0.00          -1.0        -1.0  unknown            Нет данных   

      critic_score_categories  
10435              Нет данных  
3970               Нет данных  
370                Нет данных  
6352           Средняя оценка  
2959           Средняя оценка  
...                       ...  
7104               Нет данных  
4739           Высокая оценка  
14006              Нет данных  
14033              Нет данных  
13347              Нет данных  

[12980 rows x 13 columns]

In [ ]:
# Группируем данные по платформе и находим количество продаж
count_games_platform = new_games_sort.groupby('platform')['name'].nunique()
# Выводим результат на экран отсортировав по убыванию и ограничив выдачу до 7 позиций, тем самым получив топ-7 платформ
display(count_games_platform.sort_values(ascending = False).head(7))

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1120
PS3     1086
GBA      811
Name: name, dtype: int64

Для нахождения топ-7 платформ по выпуску игр был взят период с 2000-2013г., подсчитано уникальное количество значений в столбце name, что позволило корректно выделить платформы с наибольшим количеством игр.

В результате вычислений топ-7 платформ выглядит следующим образом:
- PS2     2127
- DS      2120
- Wii     1275
- PSP     1180
- X360    1120
- PS3     1086
- GBA      811

<div class="alert alert-block alert-danger">
    
<b>Комментарий ревьюера:</b>
    
В текущем коде ты добавляешь столбец count_games, суммируя продажи игр по всем регионам, и используешь его для подсчета количества записей (строк) в группе по платформам. Однако вместо подсчета проданных копий нас просят определить топ-7 платформ по количеству уникальных игр. Для этого нужно сгруппировать данные по платформам и подсчитать количество уникальных значений в столбце name. Это позволит корректно выделить платформы с наибольшим количеством игр.
    
Кроме того, в задании сказано, что игры необходимо посчитать за актуальный период, т.е. для расчета надо использовать таблицу после фильтрации по годам.
</div>

<div class="alert alert-info">
Исправила. Для расчетов взяла данные за актуальный период с 2000-2013г., для нахождения топ-7 платформ по количеству выпущенных игр взяла данные из столбца name и посчитала их уникальное количество. Добавила выводы.
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Теперь платформы выделены верно.

</div>

---

## Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

Датасет `new_games.csv` содержит 11 столбцов и 16956 строк в которых представлена информация о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр.

Изучив первые строки таблицы, в данных были обнаружены пропуски в столбцах `Name`, `Year of Release`, `Genre`, `Critic Score`, `User Score`, `Rating`. Названия столбцов и данные в них соответствуют описанию, но названия столбцов представлены в неудобном для работы виде, поэтому названия столбцов приведены к стилю `snake case`.

Проведя анализ типов данных выявлено, что некоторые столбцы представлены некорректно, поэтому выполнили преобразование типов данных.
- Столбец `year_of_release` привели к типу `int64`
- Столбws `eu_sales`, `jp_sales`, `user_score` приведены к типу `float64`

В столбцах eu_sales, jp_sales, user_score найдены текстовые значения. Поэтому для преобразования используем метод    pd.to_numeric() с параметром errors='coerce'. Этот параметр позволяет обрабатывать ошибки, возникающие при попытке   преобразования значений, которые не могут быть корректно интерпретированы. В данном случае все значения, которые не могут быть преобразованы, будут заменены на NaN.

В данных были обнаружены пропущенные значения. Все пропуски устранены.

- В столбцах `critic_score`, `user_score` пропуски заполнили значение-индикатор `-1`.
- В столбце `rating` пропуски заменены на значение-индикатор `unknown`
- В столбце `year_of_release` пропуски удалены, чтобы избежать возможных ошибок в данных.
- В столбцах `eu_sales` и `jp_sales` пропуски заполнили средним значением.
- В столбцах `name` и `genre строк` строк с пропусками меньше 1%, поэтому пропуски были удалены.

В результате проведения предобработки данных были выявлены и преобразованы неявные дубликаты в названиях жанров игр. С помощью приведения всех названий жанров к одинаковому регистру, дубликаты устранены.

Проверили наличие дубликатов в названиях игровых платформ. Среди названий платформ все значения уникальны.

Во время поиска дубликатов в данных о рейтингах выявлено значение которого нет в расшифровках аббревиатуры рейтингов - К-А. По данным интернета этот рейтинг использовался до 1998 года, когда его переименовали в E. Значение рейтинга К-А заменили на более актаульное - Е.

Дубликатов в данных о годе выхода игр нет.

Так же в данных обнаружены и удалены явные дубликаты. Таких дубликатов 1,42%

Был выполнен срез данных за период с 2000 по 2013 год включительно, для того, чтобы изучить историю продаж игр в начале XXI века. Отберали данные по этому показателю, новый срез данных сохранили в отдельном датафрейме `df_actual`.

В датасет добавили новые поля:
- `user_score_categories` - категория оценки пользователей игр (Низкая оценка, Средняя оценка, Высокая оценка)
- `critic_score_categories` - категория оценки критиков (Низкая оценка, Средняя оценка, Высокая оценка)

Так же выделили топ-7 платформ по количеству продаж игр. В топ вошли такие платформы:
- PS2     2127 млн. проданных копий
- DS      2120 млн. проданных копий
- Wii     1275 млн. проданных копий
- PSP     1180 млн. проданных копий
- X360    1120 млн. проданных копий
- PS3     1086 млн. проданных копий
- GBA      811 млн. проданных копий

В процессе предобработки было удалено 235 строк, что составило 1,14%.

Данные полностью подготовлены.

Обновленное описание полей итогового датафрейма `/datasets/new_games.csv:`

- `name` — название игры.
- `platform` — название платформы.
- `year_of_release` — год выпуска игры.
- `genre` — жанр игры.
- `na_sales` — продажи в Северной Америке (в миллионах проданных копий).
- `eu_sales` — продажи в Европе (в миллионах проданных копий).
- `jp_sales` — продажи в Японии (в миллионах проданных копий).
- `other_sales` — продажи в других странах (в миллионах проданных копий).
- `critic_score` — оценка критиков (от 0 до 100).
- `user_score` — оценка пользователей (от 0 до 10).
- `rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.
- `user_score_categories` - категория оценки пользователей игр.
- `critic_score_categories` - категория оценки критиков.
    

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b>
    
Итоговый вывод хорошо структурирован и последовательно описывает все этапы предобработки данных, включая изменения названий столбцов, обработку пропусков, дубликатов и преобразование данных. Это говорит о проделанной тщательной работе.
</div>

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b>  
    
Можно дополнить вывод следующей информацией:
    
* указать, сколько строк было удалено в процессе предобработки;
* добавить обновлённое описание полей итогового датафрейма. Это хороший способ показать, что данные были полностью подготовлены и какие изменения внесены в структуру.
</div>

<div class="alert alert-info">
Исправила. Добавила информацию о количестве удаленных строк.
Добавила обновленное описание полей итогового датафрейма.
Скорректировала выводы после исправления и корректировок, дополнила новой информацией.
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

# <b>Итоговый комментарий ревьюера:</b>
    
<div class="alert alert-success">

Благодарю тебя за выполнение проекта. Видно, что ты постаралась, когда работала над ним. Однако есть несколько моментов, которые необходимо дополнить:
    
* вывести уникальные значения в столбцах перед заменой;
* дополнить обработку пропусков;
* скорректировать выделение топ-7 платформ.

Жду проект после доработки. Уверена, ты справишься. 💪
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Спасибо за внимательное отношение к комментариям. Проект принят, и желаю успехов в дальнейшем обучении на курсе!
    
</div>